cirq-core/cirq/transformers/routing/route_circuit_cqc.py

```diff
@@ -0,0 +1,327 @@
+            disjoint qubits. It is 'maximal' because any 2-qubit gate's qubits in the next timestep
+            must intersect with the qubits that are acted on in the current timestep.
+
+        (2) Places the logical qubits in the input circuit onto some input device by using an
+            initial mapper (LineInitialMapper by default).
+
+        (3) Insert necessary swaps to ensure all 2-qubit gates are executable on the device by
+            traversing the timesteps from left to right while:
+                (i) Removing any single qubit gate and executable 2-qubit gate in the current
+                    timestep and add it to the output routed circuit.
+               (ii) If there aren't any gates left in the current timestep, move on to the next.
+              (iii) If there are gates remaining in the current timesteps, consider a set of
+                    candidate swaps on them and rank them based on a heuristic cost function. Pick
+                    the swap that minimises the cost and use it to update our logical to physical
+                    mapping. Repeat from (i).
+    """
+
+    def __init__(self, device_graph: nx.Graph):
+        """Initializes the circuit routing transformer."""
+
+        if nx.is_directed(device_graph):
+            raise ValueError("Device graph must be undirected.")
+        self.device_graph = device_graph
+
+    def __call__(
+        self,
+        circuit: 'cirq.AbstractCircuit',
+        *,
+        max_search_radius: int = 8,
+        preserve_moment_structure: bool = False,
+        tag_inserted_swaps: bool = False,
+        initial_mapper: Optional['cirq.AbstractInitialMapper'] = None,
+        context: Optional['cirq.TransformerContext'] = None,
+    ) -> 'cirq.AbstractCircuit':
+        """Transforms the given circuit to make it executable on the device.
+
+        Since routing doesn't necessarily modify any specific operation and only adds swaps
+        before /after operations to ensure the circuit can be executed, tagging operations with
+        tags from context.tags_to_ignore will have no impact on the routing procedure.
+
+        Args:
+            circuit: the input circuit to be transformed.
+            max_search_radius: the maximum number of times the cost function can be iterated for
+                convergence.
+            preverse_moment_structure: whether or not the transfomer should preserve the given
+                moment structure of 'circuit'.
+            tag_inserted_swaps: whether or not a RoutingSwapTag should be attched to inserted swap
+                operations.
+            initial_mapper: an initial mapping strategy (placement) of logical qubits in the
+                circuit onto physical qubits on the device.
+            context: transformer context storing common configurable options for transformers.
+
+        Returns:
+            The routed circuit executable on the harware with the same unitary as 'circuit'.
```

Comment:
1. Routed circuit might not still be executable on hardware because of a different gateset (you are also not expecting a device in the constructor anymore)
2. The circuit might not be a "unitary" (eg: containing measurements + classical feedback)
```suggestion
            The routed circuit, which is equivalent to original circuit upto a final qubit permutation and where each 2-qubit operation is between adjacent qubits in the `device_graph`. 
```